pkg/apis/k3os.io/v1/types.go

```diff
@@ -0,0 +1,116 @@
+package v1
+
+// Copyright 2019 Rancher Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import (
+	"time"
+
+	"github.com/rancher/wrangler/pkg/condition"
+	"github.com/rancher/wrangler/pkg/genericcondition"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	// ChannelLatestResolved indicates the value for latest has been polled and resolved
+	ChannelLatestResolved = condition.Cond("LatestResolved")
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Channel represents a series of releases. The only protocol that a Channel
+// must adhere to is GETing the URL will return a 302 with the last segment
+// of the Location being the "latest" tag/release.
+type Channel struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ChannelSpec   `json:"spec"`
+	Status ChannelStatus `json:"status"`
+}
+
+// ChannelSpec represents the user-configurable details of a Channel
+type ChannelSpec struct {
+	URL string `json:"url,omitempty"` // e.g. https://github.com/rancher/k3os/releases/latest
+}
+
+// ChannelStatus represents the resulting state from processing Channel events.
+type ChannelStatus struct {
+	LatestVersion string                              `json:"latestVersion,omitempty"`
+	Conditions    []genericcondition.GenericCondition `json:"conditions,omitempty"`
+}
+
+type DrainSpec struct {
+	Timeout          *time.Duration `json:"timeout,omitempty"`
+	GracePeriod      *int32         `json:"gracePeriod,omitempty"`
+	DeleteLocalData  bool           `json:"deleteLocalData,omitempty"`
```

Comment:
I'd change to `*bool` and nil means "true".  Same for IgnoreDaemonSets.  I think most of the time that is what most admins would want.